In [21]:
#Task 1
import pandas as pd
df=pd.read_csv("adult.csv")
df

,age,workclass,fnlwgt,education,education.num,marital.status,occupation,relationship,race,sex,capital.gain,capital.loss,hours.per.week,native.country,income
0,90,?,77053,HS-grad,9,Widowed,?,Not-in-family,White,Female,0,4356,40,United-States,<=50K
1,82,Private,132870,HS-grad,9,Widowed,Exec-managerial,Not-in-family,White,Female,0,4356,18,United-States,<=50K
2,66,?,186061,Some-college,10,Widowed,?,Unmarried,Black,Female,0,4356,40,United-States,<=50K
3,54,Private,140359,7th-8th,4,Divorced,Machine-op-inspct,Unmarried,White,Female,0,3900,40,United-States,<=50K
4,41,Private,264663,Some-college,10,Separated,Prof-specialty,Own-child,White,Female,0,3900,40,United-States,<=50K
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
32556,22,Private,310152,Some-college,10,Never-married,Protective-serv,Not-in-family,White,Male,0,0,40,United-States,<=50K
32557,27,Private,257302,Assoc-acdm,12,Married-civ-spouse,Tech-support,Wife,White,Female,0,0,38,United-States,<=50K
32558,40,Private,154374,HS-grad,9,Married-civ-spouse,Machine-op-inspct,Husband,White,Male,0,0,40,United-States,>50K
32559,58,Private,151910,HS-grad,9,Widowed,Adm-clerical,Unmarried,White,Female,0,0,40,United-States,<=50K


In [22]:
import numpy as np
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32561 entries, 0 to 32560
Data columns (total 15 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   age             32561 non-null  int64 
 1   workclass       32561 non-null  object
 2   fnlwgt          32561 non-null  int64 
 3   education       32561 non-null  object
 4   education.num   32561 non-null  int64 
 5   marital.status  32561 non-null  object
 6   occupation      32561 non-null  object
 7   relationship    32561 non-null  object
 8   race            32561 non-null  object
 9   sex             32561 non-null  object
 10  capital.gain    32561 non-null  int64 
 11  capital.loss    32561 non-null  int64 
 12  hours.per.week  32561 non-null  int64 
 13  native.country  32561 non-null  object
 14  income          32561 non-null  object
dtypes: int64(6), object(9)
memory usage: 3.7+ MB


In [24]:
import numpy as np


print("Missing values before replacing ? ")
display(df.isnull().sum())

df.replace('?', np.nan, inplace=True)
print("Unique values in 'workclass' before replacement:")


print('\nMissing values after replacing "?" with NaN:')
display(df.isnull().sum())

print('\nHead of the DataFrame after replacement:')
display(df.head())

Missing values before replacing ? 


,0
age,0
workclass,0
fnlwgt,0
education,0
education.num,0
marital.status,0
occupation,0
relationship,0
race,0
sex,0


Unique values in 'workclass' before replacement:

Missing values after replacing "?" with NaN:


,0
age,0
workclass,1836
fnlwgt,0
education,0
education.num,0
marital.status,0
occupation,1843
relationship,0
race,0
sex,0



Head of the DataFrame after replacement:


,age,workclass,fnlwgt,education,education.num,marital.status,occupation,relationship,race,sex,capital.gain,capital.loss,hours.per.week,native.country,income
0,90,NaN,77053,HS-grad,9,Widowed,NaN,Not-in-family,White,Female,0,4356,40,United-States,<=50K
1,82,Private,132870,HS-grad,9,Widowed,Exec-managerial,Not-in-family,White,Female,0,4356,18,United-States,<=50K
2,66,NaN,186061,Some-college,10,Widowed,NaN,Unmarried,Black,Female,0,4356,40,United-States,<=50K
3,54,Private,140359,7th-8th,4,Divorced,Machine-op-inspct,Unmarried,White,Female,0,3900,40,United-States,<=50K
4,41,Private,264663,Some-college,10,Separated,Prof-specialty,Own-child,White,Female,0,3900,40,United-States,<=50K


In [27]:
#Task 2
missing_counts = df.isnull().sum()
missing_percentages = (df.isnull().sum() / len(df)) * 100

missing_info = pd.DataFrame({
    'Missing Count': missing_counts,
    'Missing Percentage': missing_percentages
})

missing_info = missing_info[missing_info['Missing Count'] > 0].sort_values(by='Missing Count', ascending=False)

display(missing_info)


,Missing Count,Missing Percentage
occupation,1843,5.660146
workclass,1836,5.638647
native.country,583,1.790486


In [33]:
#Task 3
for col in ['workclass', 'occupation']:
    common_value = df[col].mode()[0]
    df[col].fillna(common_value, inplace=True)

print("Missing values after imputation:")
display(df.isnull().sum())

print("\nHead of the DataFrame after imputation:")
display(df.head())

Missing values after imputation:


,0
age,0
workclass,0
fnlwgt,0
education,0
education.num,0
marital.status,0
occupation,0
relationship,0
race,0
sex,0



Head of the DataFrame after imputation:


,age,workclass,fnlwgt,education,education.num,marital.status,occupation,relationship,race,sex,capital.gain,capital.loss,hours.per.week,native.country,income
0,90,Private,77053,HS-grad,9,Widowed,Prof-specialty,Not-in-family,White,Female,0,4356,40,United-States,<=50K
1,82,Private,132870,HS-grad,9,Widowed,Exec-managerial,Not-in-family,White,Female,0,4356,18,United-States,<=50K
2,66,Private,186061,Some-college,10,Widowed,Prof-specialty,Unmarried,Black,Female,0,4356,40,United-States,<=50K
3,54,Private,140359,7th-8th,4,Divorced,Machine-op-inspct,Unmarried,White,Female,0,3900,40,United-States,<=50K
4,41,Private,264663,Some-college,10,Separated,Prof-specialty,Own-child,White,Female,0,3900,40,United-States,<=50K


In [46]:
#Task 4
print('Checking for duplicated rows')
print(df.duplicated().sum())

Checking for duplicated rows
24


In [57]:
#Task 5
for col in ['education','marital.status','native.country']:
    print(f"Unique values for {col}:")
    df[col] = df[col].astype(str).str.strip()
    print(df[col].unique())
    print("\n")

Unique values for education:
['HS-grad' 'Some-college' '7th-8th' '10th' 'Doctorate' 'Prof-school'
 'Bachelors' 'Masters' '11th' 'Assoc-acdm' 'Assoc-voc' '1st-4th' '5th-6th'
 '12th' '9th' 'Preschool']


Unique values for marital.status:
['Widowed' 'Divorced' 'Separated' 'Never-married' 'Married-civ-spouse'
 'Married-spouse-absent' 'Married-AF-spouse']


Unique values for native.country:
['United-States' 'nan' 'Mexico' 'Greece' 'Vietnam' 'China' 'Taiwan'
 'India' 'Philippines' 'Trinadad&Tobago' 'Canada' 'South'
 'Holand-Netherlands' 'Puerto-Rico' 'Poland' 'Iran' 'England' 'Germany'
 'Italy' 'Japan' 'Hong' 'Honduras' 'Cuba' 'Ireland' 'Cambodia' 'Peru'
 'Nicaragua' 'Dominican-Republic' 'Haiti' 'El-Salvador' 'Hungary'
 'Columbia' 'Guatemala' 'Jamaica' 'Ecuador' 'France' 'Yugoslavia'
 'Scotland' 'Portugal' 'Laos' 'Thailand' 'Outlying-US(Guam-USVI-etc)']




In [58]:
#Task 6
print("\nDescriptive Statistics for Numeric Columns:")
display(df.describe().T)

print("\nDescriptive Statistics for Categorical Columns:")
display(df.describe(include='object').T)

print("\nAnalysis of 'age' and 'hours.per.week':")

age_mean = df['age'].mean()
age_median = df['age'].median()
hours_mean = df['hours.per.week'].mean()
hours_median = df['hours.per.week'].median()
print(f"Mean of 'age': {age_mean:.2f}")
print(f"Median of 'age': {age_median}")
print(f"Mean of 'hours.per.week': {hours_mean:.2f}")
print(f"Median of 'hours.per.week': {hours_median}")


Descriptive Statistics for Numeric Columns:


,count,mean,std,min,25%,50%,75%,max
age,32561.0,38.581647,13.640433,17.0,28.0,37.0,48.0,90.0
fnlwgt,32561.0,189778.366512,105549.977697,12285.0,117827.0,178356.0,237051.0,1484705.0
education.num,32561.0,10.080679,2.572720,1.0,9.0,10.0,12.0,16.0
capital.gain,32561.0,1077.648844,7385.292085,0.0,0.0,0.0,0.0,99999.0
capital.loss,32561.0,87.303830,402.960219,0.0,0.0,0.0,0.0,4356.0
hours.per.week,32561.0,40.437456,12.347429,1.0,40.0,40.0,45.0,99.0



Descriptive Statistics for Categorical Columns:


,count,unique,top,freq
workclass,32561,8,Private,24532
education,32561,16,HS-grad,10501
marital.status,32561,7,Married-civ-spouse,14976
occupation,32561,14,Prof-specialty,5983
relationship,32561,6,Husband,13193
race,32561,5,White,27816
sex,32561,2,Male,21790
native.country,32561,42,United-States,29170
income,32561,2,<=50K,24720



Analysis of 'age' and 'hours.per.week':
Mean of 'age': 38.58
Median of 'age': 37.0
Mean of 'hours.per.week': 40.44
Median of 'hours.per.week': 40.0


The mean and median of both columns are approximately equal which mean that there are not many outliers.

In [59]:
#Task 7
print("\n--- Task 7: Frequency Analysis ---\n")

print(f"Most common occupation: {df['occupation'].mode()[0]}")

print("\nPercentage distribution of sex:")
print(df['sex'].value_counts(normalize=True) * 100)

print("\nPercentage distribution of income label:")
print(df['income'].value_counts(normalize=True) * 100)


--- Task 7: Frequency Analysis ---

Most common occupation: Prof-specialty

Percentage distribution of sex:
sex
Male      66.920549
Female    33.079451
Name: proportion, dtype: float64

Percentage distribution of income label:
income
<=50K    75.919044
>50K     24.080956
Name: proportion, dtype: float64


#Comment on income classes
most of the people lie below 50k.Almost 75% of the total people which means the income is not well balanced

In [60]:
print("\n--- Task 8: Cross-Check a Data Quality Assumption ---\n")

education_consistency = df.groupby('education')['education.num'].unique()
print("Consistency check for 'education' and 'education.num':")
print(education_consistency)
inconsistencies = education_consistency[education_consistency.apply(len) > 1]
print(inconsistencies)
print("\nReport on inconsistencies:")
if not inconsistencies.empty:
    print("Inconsistencies found: Each education level should map to a single education.num, but the following show multiple mappings:")
    print(inconsistencies)
else:
    print("No inconsistencies found. Each education level maps consistently to a single education.num.")


--- Task 8: Cross-Check a Data Quality Assumption ---

Consistency check for 'education' and 'education.num':
education
10th             [6]
11th             [7]
12th             [8]
1st-4th          [2]
5th-6th          [3]
7th-8th          [4]
9th              [5]
Assoc-acdm      [12]
Assoc-voc       [11]
Bachelors       [13]
Doctorate       [16]
HS-grad          [9]
Masters         [14]
Preschool        [1]
Prof-school     [15]
Some-college    [10]
Name: education.num, dtype: object
Series([], Name: education.num, dtype: object)

Report on inconsistencies:
No inconsistencies found. Each education level maps consistently to a single education.num.


In [64]:
#Task 9
import sqlite3
import pandas as pd
csv_file = "adult.csv"

original_df = pd.read_csv(csv_file)
print(f"Original CSV Shape: {original_df.shape}")

conn = sqlite3.connect(":memory:")

original_df.to_sql("adult", conn, if_exists="replace", index=False)

sql_query = "SELECT * FROM adult WHERE age > 30"
sql_filtered_df = pd.read_sql_query(sql_query, conn)

csv_filtered_df = original_df[original_df["age"] > 30]

print(f"DataFrame filtered from CSV shape: {csv_filtered_df.shape}")
print(f"DataFrame extracted via SQL shape: {sql_filtered_df.shape}")

if csv_filtered_df.shape == sql_filtered_df.shape:
    print("Verification Successful: Shapes match exactly!")
else:
    print("Verification Failed: Shapes do not match.")

conn.close()

Original CSV Shape: (32561, 15)
DataFrame filtered from CSV shape: (21989, 15)
DataFrame extracted via SQL shape: (21989, 15)
Verification Successful: Shapes match exactly!


## Data Profiling Summary Report

### 1. Dataset Overview

The dataset contains information about individuals to predict income levels. It has **32,561 rows** and **15 columns**, including demographic information (age, sex, marital status, race, native country), educational background (education, education.num), employment details (workclass, occupation, hours.per.week), financial aspects (capital.gain, capital.loss, fnlwgt), and the target variable, 'income'.

### 2. Data Quality Issues Found

*   **Missing Values**: Initially, values represented by '?' were identified. These were replaced with nan values. The columns with missing values and their counts/percentages are:
    *   `occupation`: 1843 missing values (5.66%)
    *   `workclass`: 1836 missing values (5.64%)
    *   `native.country`: 583 missing values (1.79%)

*   **Duplicates**: **24 duplicate rows** were found in the dataset.

*   **Inconsistent Data**: consistency between 'education' and 'education.num'. No inconsistencies were found.

### 3. Key Observations

*   **Income Class Imbalance**: The 'income' distribution is significantly imbalanced:
    *   `<=50K`: 75.92%
    *   `>50K`: 24.08%
    

*   **Most Common Occupation**: The most common occupation in the dataset is **'Prof-specialty'**.

*   **Sex Distribution**: The dataset consists of approximately `66.92% Male` and `33.08% Female` individuals.

*   **Age and Hours per Week**: The mean and median for 'age' (38.58 and 37.0, respectively) and 'hours.per.week' (40.44 and 40.0, respectively) are very close.


### 4. Cleaning Decisions Made

*   **Missing Value Handling**: The '?' character, indicating missing data, was replaced with nan across the DataFrame.

*   **Imputation**: For the 'workclass' and 'occupation' columns, missing values were imputed using the **mode** (most frequent value) of their respective columns. This approach was chosen to preserve the categorical nature of these features.
*   **Whitespace Stripping**: For categorical columns like 'education', 'marital.status', and 'native.country', leading/trailing whitespaces were removed using .str.strip() to ensure consistency and accurate unique values.